File: solutions/c/space-age/3/space_age.c
#include "space_age.h"

static const double orbital_period[] = {0.2408467, 0.61519726, 1,
                                        1.8808158, 11.862615,  29.447498,
                                        84.016846, 164.79132};

float age(planet_t planet, int64_t seconds) {
  const int earth_year = 31557600;

  if (planet > NEPTUNE)
    return INVALID_PLANET;
  return seconds / (earth_year * orbital_period[planet]);
}

File: solutions/haskell/space-age/3/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthYear = 31557600

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / year
  where
    year = case planet of
      Earth -> earthYear
      Mercury -> earthYear * 0.2408467
      Venus -> earthYear * 0.61519726
      Mars -> earthYear * 1.8808158
      Jupiter -> earthYear * 11.862615
      Saturn -> earthYear * 29.447498
      Uranus -> earthYear * 84.016846
      Neptune -> earthYear * 164.79132